File: python/gz/proto/hello.py
from __future__ import annotations

import struct
from dataclasses import dataclass

from gz.common.tags import (
    ActionSetHash,
    EngineIdentity,
    EngineId,
    EngineVersion,
    FeatureSchemaHash,
    ModelVersion,
)
from gz.proto.errors import ERROR_MALFORMED, ProtocolError

HELLO_LEN = 108
HELLO_ACK_LEN = 28


@dataclass(frozen=True, slots=True)
class Hello:
    protocol_version: int
    encoding_version: int
    feature_schema_hash: FeatureSchemaHash
    batch_capacity: int
    engine_identity: EngineIdentity

    @property
    def engine_id(self) -> EngineId:
        return self.engine_identity.engine_id

    @property
    def engine_version(self) -> EngineVersion:
        return self.engine_identity.engine_version

    @property
    def action_set_hash(self) -> ActionSetHash:
        return self.engine_identity.action_set_hash

    def encode(self) -> bytes:
        return (
            struct.pack("<II", self.protocol_version, self.encoding_version)
            + bytes(self.feature_schema_hash)
            + struct.pack("<I", self.batch_capacity)
            + bytes(self.engine_id)
            + bytes(self.engine_version)
            + bytes(self.action_set_hash)
        )

    @classmethod
    def decode(cls, buf: memoryview) -> Hello:
        if len(buf) != HELLO_LEN:
            raise ProtocolError(ERROR_MALFORMED, "bad hello length")
        protocol_version, encoding_version = struct.unpack_from("<II", buf, 0)
        try:
            engine_identity = EngineIdentity.from_parts(
                EngineId.from_bytes(buf[44:60]),
                EngineVersion.from_bytes(buf[60:76]),
                ActionSetHash.from_bytes(buf[76:108]),
            )
        except ValueError as error:
            raise ProtocolError(ERROR_MALFORMED, str(error)) from error
        return cls(
            protocol_version=protocol_version,
            encoding_version=encoding_version,
            feature_schema_hash=FeatureSchemaHash.from_bytes(buf[8:40]),
            batch_capacity=struct.unpack_from("<I", buf, 40)[0],
            engine_identity=engine_identity,
        )


@dataclass(frozen=True, slots=True)
class HelloAck:
    protocol_version: int
    model_version: ModelVersion
    model_generation: int

    def encode(self) -> bytes:
        return (
            struct.pack("<I", self.protocol_version)
            + bytes(self.model_version)
            + struct.pack("<Q", self.model_generation)
        )

    @classmethod
    def decode(cls, buf: memoryview) -> HelloAck:
        if len(buf) != HELLO_ACK_LEN:
            raise ProtocolError(ERROR_MALFORMED, "bad hello ack length")
        return cls(
            protocol_version=struct.unpack_from("<I", buf, 0)[0],
            model_version=ModelVersion.from_bytes(buf[4:20]),
            model_generation=struct.unpack_from("<Q", buf, 20)[0],
        )

File: crates/gz-eval-service/src/hello.rs
use crate::{PROTOCOL_VERSION, ServiceError, ServiceResult};
use gz_engine::{ActionSetHash, EngineId, EngineVersion, ModelVersion};
use gz_features::{BATCH_ENCODING_VERSION, FeatureSchemaHash};

pub const ERROR_PROTOCOL: u32 = 1;
pub const ERROR_ENCODING: u32 = 2;
pub const ERROR_SCHEMA: u32 = 3;
pub const ERROR_CAPACITY: u32 = 4;
pub const ERROR_MALFORMED: u32 = 5;

const HELLO_LEN: usize = 108;
const HELLO_ACK_LEN: usize = 28;
const ERROR_HEADER_LEN: usize = 6;
const MAX_ERROR_MESSAGE: usize = 512;

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct Hello {
    pub protocol_version: u32,
    pub encoding_version: u32,
    pub feature_schema_hash: FeatureSchemaHash,
    pub batch_capacity: u32,
    pub engine_id: EngineId,
    pub engine_version: EngineVersion,
    pub action_set_hash: ActionSetHash,
}

impl Hello {
    #[must_use]
    pub const fn new(
        feature_schema_hash: FeatureSchemaHash,
        batch_capacity: u32,
        engine_id: EngineId,
        engine_version: EngineVersion,
        action_set_hash: ActionSetHash,
    ) -> Self {
        Self {
            protocol_version: PROTOCOL_VERSION,
            encoding_version: BATCH_ENCODING_VERSION,
            feature_schema_hash,
            batch_capacity,
            engine_id,
            engine_version,
            action_set_hash,
        }
    }

    pub fn encode(&self, out: &mut Vec<u8>) {
        out.clear();
        out.extend_from_slice(&self.protocol_version.to_le_bytes());
        out.extend_from_slice(&self.encoding_version.to_le_bytes());
        out.extend_from_slice(self.feature_schema_hash.as_bytes());
        out.extend_from_slice(&self.batch_capacity.to_le_bytes());
        out.extend_from_slice(self.engine_id.as_bytes());
        out.extend_from_slice(self.engine_version.as_bytes());
        out.extend_from_slice(self.action_set_hash.as_bytes());
    }

    pub fn decode(bytes: &[u8]) -> ServiceResult<Self> {
        if bytes.len() != HELLO_LEN {
            return Err(ServiceError::handshake("bad HELLO length"));
        }
        Ok(Self {
            protocol_version: read_u32(bytes, 0)?,
            encoding_version: read_u32(bytes, 4)?,
            feature_schema_hash: FeatureSchemaHash::from_bytes(read_array(bytes, 8)?),
            batch_capacity: read_u32(bytes, 40)?,
            engine_id: EngineId::from_bytes(read_array(bytes, 44)?),
            engine_version: EngineVersion::from_bytes(read_array(bytes, 60)?),
            action_set_hash: ActionSetHash::from_bytes(read_array(bytes, 76)?),
        })
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct HelloAck {
    pub protocol_version: u32,
    pub model_version: ModelVersion,
    pub model_generation: u64,
}

impl HelloAck {
    pub fn encode(&self, out: &mut Vec<u8>) {
        out.clear();
        out.extend_from_slice(&self.protocol_version.to_le_bytes());
        out.extend_from_slice(self.model_version.as_bytes());
        out.extend_from_slice(&self.model_generation.to_le_bytes());
    }

    pub fn decode(bytes: &[u8]) -> ServiceResult<Self> {
        if bytes.len() != HELLO_ACK_LEN {
            return Err(ServiceError::handshake("bad HELLO_ACK length"));
        }
        Ok(Self {
            protocol_version: read_u32(bytes, 0)?,
            model_version: ModelVersion::from_bytes(read_array(bytes, 4)?),
            model_generation: u64::from_le_bytes(read_array(bytes, 20)?),
        })
    }
}

pub fn decode_error(bytes: &[u8]) -> ServiceResult<(u32, String)> {
    if bytes.len() < ERROR_HEADER_LEN {
        return Err(ServiceError::protocol("ERROR frame truncated"));
    }
    let code = read_u32(bytes, 0)?;
    let message_len = u16::from_le_bytes(
        bytes[4..6]
            .try_into()
            .expect("error message length slice checked"),
    ) as usize;
    if message_len > MAX_ERROR_MESSAGE || bytes.len() != ERROR_HEADER_LEN + message_len {
        return Err(ServiceError::protocol("bad ERROR frame length"));
    }
    let message = String::from_utf8_lossy(&bytes[ERROR_HEADER_LEN..]).into_owned();
    Ok((code, message))
}

fn read_u32(bytes: &[u8], offset: usize) -> ServiceResult<u32> {
    let slice = bytes
        .get(offset..offset + 4)
        .ok_or_else(|| ServiceError::protocol("u32 truncated"))?;
    Ok(u32::from_le_bytes(slice.try_into().expect("slice checked")))
}

fn read_array<const N: usize>(bytes: &[u8], offset: usize) -> ServiceResult<[u8; N]> {
    let slice = bytes
        .get(offset..offset + N)
        .ok_or_else(|| ServiceError::protocol("byte array truncated"))?;
    Ok(slice.try_into().expect("slice checked"))
}
